task-impls/src/da.rs

```diff
@@ -0,0 +1,334 @@
+                vote.vote_data,
+                vote.vote_token.clone(),
+                state.cur_view,
+                accumulator,
+                None,
+            ) {
+                Left(acc) => {
+                    state.accumulator = Either::Left(acc);
+                    return (None, state);
+                }
+                Right(dac) => {
+                    state
+                        .event_stream
+                        .publish(SequencingHotShotEvent::DACSend(dac.clone()))
+                        .await;
+                    state.accumulator = Right(dac);
+                    return (None, state);
+                }
+            }
+        }
+        SequencingHotShotEvent::Shutdown => return (Some(HotShotTaskCompleted::ShutDown), state),
+        _ => {}
+    }
+    (None, state)
+}
+
+impl<
+        TYPES: NodeType<ConsensusType = SequencingConsensus, Time = ViewNumber>,
+        I: NodeImplementation<
+            TYPES,
+            Leaf = SequencingLeaf<TYPES>,
+            ConsensusMessage = SequencingMessage<TYPES, I>,
+        >,
+    > DATaskState<TYPES, I>
+where
+    I::Exchanges: SequencingExchangesType<TYPES, Message<TYPES, I>>,
+    CommitteeEx<TYPES, I>: ConsensusExchange<
+        TYPES,
+        Message<TYPES, I>,
+        Certificate = DACertificate<TYPES>,
+        Commitment = TYPES::BlockType,
+    >,
+{
+    pub async fn handle_event(
+        &mut self,
+        event: SequencingHotShotEvent<TYPES, I>,
+    ) -> Option<HotShotTaskCompleted> {
+        match event {
+            SequencingHotShotEvent::DAProposalRecv(proposal, sender) => {
+                let view = proposal.data.get_view_number();
+                if view < self.cur_view {
+                    return None;
+                }
+                let block_commitment = proposal.data.deltas.commit();
+                let view_leader_key = self.committee_exchange.get_leader(view);
+                if view_leader_key != sender {
+                    return None;
+                }
+                if !view_leader_key.validate(&proposal.signature, block_commitment.as_ref()) {
+                    warn!(?proposal.signature, "Could not verify proposal.");
+                    return None;
+                }
+
+                let vote_token = self.committee_exchange.make_vote_token(view);
+                match vote_token {
+                    Err(e) => {
+                        error!("Failed to generate vote token for {:?} {:?}", view, e);
+                    }
+                    Ok(None) => {
+                        info!("We were not chosen for DA committee on {:?}", view);
+                    }
+                    Ok(Some(vote_token)) => {
+                        info!("We were chosen for DA committee on {:?}", view);
+
+                        // Generate and send vote
+                        let message = self.committee_exchange.create_da_message::<I>(
+                            self.high_qc.commit(),
```

Comment:
Looks like this never actually gets updated.  Maybe the consensus task can emit an event when it updates it's high_qc, or maybe this qc should come from the DA proposal?  TBH I'm not really sure if we need a QC on the DA message now that it's separated from the consensus logic.